### PR TITLE
parallels: update support for Parallels Tools

### PR DESCRIPTION
**Pull request by paulsmith**
Install Parallel Tools updated for version 17 of Parallels for macOS. This
fixes clipboard sharing, so that copy and paste works between the host
macOS and the guest NixOS VM. Support for guests on M1 Apple Silicon-based
Macs (aarch64-linux) is also added.

### CODE REVIEW

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 1]
```diff
@@ -57,109 +64,56 @@ stdenv.mkDerivation rec {
...
   installPhase = ''
     if test -z "$libsOnly"; then
       ( # kernel modules
         cd kmods
         mkdir -p $out/lib/modules/${kernelVersion}/extra
-        cp prl_eth/pvmnet/prl_eth.ko $out/lib/modules/${kernelVersion}/extra
-        cp prl_tg/Toolgate/Guest/Linux/prl_tg/prl_tg.ko $out/lib/modules/${kernelVersion}/extra
         cp prl_fs/SharedFolders/Guest/Linux/prl_fs/prl_fs.ko $out/lib/modules/${kernelVersion}/extra
         cp prl_fs_freeze/Snapshot/Guest/Linux/prl_freeze/prl_fs_freeze.ko $out/lib/modules/${kernelVersion}/extra
+        cp prl_notifier/Installation/lnx/prl_notifier/prl_notifier.ko $out/lib/modules/${kernelVersion}/extra
```
Comment (wegank) [inline pkgs/os-specific/linux/prl-tools/default.nix:77]: prl_notifier should only be built for aarch64.

File: nixos/modules/virtualisation/parallels-guest.nix [hunk 2]
```diff
@@ -45,34 +46,36 @@ in
...
+    hardware.opengl.extraPackages32 = [ pkg.pkgsi686Linux.mesa.drivers ];
 
     services.udev.packages = [ prl-tools ];
 
     environment.systemPackages = [ prl-tools ];
 
     boot.extraModulePackages = [ prl-tools ];
 
-    boot.kernelModules = [ "prl_tg" "prl_eth" "prl_fs" "prl_fs_freeze" ];
+    boot.kernelModules = [ "prl_fs" "prl_fs_freeze" "prl_notifier" "prl_tg" ];
```
Comment (wegank) [inline nixos/modules/virtualisation/parallels-guest.nix:78]: prl_notifier should only be built for aarch64.

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 3]
```diff
@@ -57,109 +64,56 @@ stdenv.mkDerivation rec {
...
-      cd $out
-      find -name \*.so\* -type f -exec \
-        patchelf --set-rpath "$out/lib:$libPath" {} \;
-
-      cd lib
-      libGLname=$(echo libGL.so*)
-      ln -s $libGLname libGL.so
-      ln -s $libGLname libGL.so.1
+      cd $out/lib
+      ln -s libPrlWl.so.1.0.0 libPrlWl.so.1
```
Comment (wegank) [inline pkgs/os-specific/linux/prl-tools/default.nix:109]: Adding these lines fixes the build issue for me.

```
      ${if aarch64 then "" else "
      ln -s libGL.so.1.0.0 libGL.so
      ln -s libGL.so.1.0.0 libGL.so.1
      ln -s libPrlDRI.so.1.0.0 libPrlDRI.so.1
      ln -s libEGL.so.1.0.0 libEGL.so.1
      ln -s libgbm.so.1.0.0 libgbm.so.1
      "}
```

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 4]
```diff
@@ -1,50 +1,57 @@
 { stdenv, lib, makeWrapper, p7zip
-, gawk, util-linux, xorg, glib, dbus-glib, zlib
+, gawk, utillinux, xorg, glib, dbus-glib, zlib
```
Comment (wegank) [inline pkgs/os-specific/linux/prl-tools/default.nix:2]: ```utillinux``` -> ```util-linux```

File: nixos/modules/virtualisation/parallels-guest.nix [hunk 5]
```diff
@@ -45,34 +47,36 @@ in
...
+
+      screenSection = ''
+        Option "NoMTRR"
+      '';
     };
 
     hardware.opengl.package = prl-tools;
     hardware.opengl.package32 = pkgs.pkgsi686Linux.linuxPackages.prl-tools.override { libsOnly = true; kernel = null; };
-    hardware.opengl.setLdLibraryPath = true;
+    hardware.opengl.extraPackages = [ pkgs.mesa.drivers ];
```
Comment (abbradar) [inline nixos/modules/virtualisation/parallels-guest.nix:70]: Why are they needed when we use Parallels driver?

File: nixos/modules/virtualisation/parallels-guest.nix [hunk 6]
```diff
@@ -45,34 +47,36 @@ in
...
+    hardware.opengl.extraPackages32 = [ pkg.pkgsi686Linux.mesa.drivers ];
 
     services.udev.packages = [ prl-tools ];
 
     environment.systemPackages = [ prl-tools ];
 
     boot.extraModulePackages = [ prl-tools ];
 
-    boot.kernelModules = [ "prl_tg" "prl_eth" "prl_fs" "prl_fs_freeze" ];
+    boot.kernelModules = if aarch64 then [ "prl_fs" "prl_fs_freeze" "prl_notifier" "prl_tg" ] else [ "prl_fs" "prl_fs_freeze" "prl_tg" ];
```
Comment (abbradar) [inline nixos/modules/virtualisation/parallels-guest.nix:79]: Better do `[ "prl_fs" "prl_fs_freeze" "prl_tg" ] ++ optional aarch64 "prl_notifier"`.

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 7]
```diff
@@ -57,109 +64,63 @@ stdenv.mkDerivation rec {
...
-      find -name \*.so\* -type f -exec \
-        patchelf --set-rpath "$out/lib:$libPath" {} \;
-
-      cd lib
-      libGLname=$(echo libGL.so*)
-      ln -s $libGLname libGL.so
-      ln -s $libGLname libGL.so.1
+      cd $out/lib
+      ln -s libPrlWl.so.1.0.0 libPrlWl.so.1
+      ${if aarch64 then "" else "
```
Comment (abbradar) [inline pkgs/os-specific/linux/prl-tools/default.nix:110]: Usual pattern is `${optionalString aarch64 ''...''}`

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 8]
```diff
@@ -57,109 +64,63 @@ stdenv.mkDerivation rec {
...
   installPhase = ''
     if test -z "$libsOnly"; then
       ( # kernel modules
         cd kmods
         mkdir -p $out/lib/modules/${kernelVersion}/extra
-        cp prl_eth/pvmnet/prl_eth.ko $out/lib/modules/${kernelVersion}/extra
-        cp prl_tg/Toolgate/Guest/Linux/prl_tg/prl_tg.ko $out/lib/modules/${kernelVersion}/extra
         cp prl_fs/SharedFolders/Guest/Linux/prl_fs/prl_fs.ko $out/lib/modules/${kernelVersion}/extra
         cp prl_fs_freeze/Snapshot/Guest/Linux/prl_freeze/prl_fs_freeze.ko $out/lib/modules/${kernelVersion}/extra
+        ${if aarch64 then "cp prl_notifier/Installation/lnx/prl_notifier/prl_notifier.ko $out/lib/modules/${kernelVersion}/extra" else ""}
```
Comment (abbradar) [inline pkgs/os-specific/linux/prl-tools/default.nix:77]: Use `optionalString` here too.

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 9]
```diff
@@ -1,48 +1,55 @@
...
-let xorgFullVer = lib.getVersion xorg.xorgserver;
-    xorgVer = lib.versions.majorMinor xorgFullVer;
-    x64 = if stdenv.hostPlatform.system == "x86_64-linux" then true
-          else if stdenv.hostPlatform.system == "i686-linux" then false
-          else throw "Parallels Tools for Linux only support {x86-64,i686}-linux targets";
+let
+  aarch64 = stdenv.hostPlatform.system == "aarch64-linux";
+  x86_64 = stdenv.hostPlatform.system == "x86_64-linux";
+  i686 = stdenv.hostPlatform.system == "i686-linux";
+  _ = if (aarch64 || x86_64 || i686) then true
```
Comment (abbradar) [inline pkgs/os-specific/linux/prl-tools/default.nix:13]: This won't fire at all, because the language is lazy. You could use `assert aarch64 || x86_64 || i686` after `in` instead.

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 10]
```diff
@@ -1,48 +1,55 @@
...
     undmg "${src}"
 
     export sourceRoot=prl-tools-build
-    7z x "Parallels Desktop.app/Contents/Resources/Tools/prl-tools-lin.iso" -o$sourceRoot
+    7z x "Parallels Desktop.app/Contents/Resources/Tools/prl-tools-lin${if aarch64 then "-arm" else ""}.iso" -o$sourceRoot
     if test -z "$libsOnly"; then
       ( cd $sourceRoot/kmods; tar -xaf prl_mod.tar.gz )
     fi
-    ( cd $sourceRoot/tools; tar -xaf prltools${if x64 then ".x64" else ""}.tar.gz )
+    # TODO
```
Comment (abbradar) [inline pkgs/os-specific/linux/prl-tools/default.nix:48]: What remains to be done here? Also I guess we want to drop the comment below -- it's not relevant anymore, or is it?

**Review by abbradar — COMMENTED** [review]
Thanks for making this into a PR!

I left myself in `maintainers` before to be able to explain how the package works, even though I couldn't actually update it. The package right now is even more unknown to me, so let's drop myself from maintainers.

File: pkgs/os-specific/linux/prl-tools/default.nix (hunk 7 above)
Comment (abbradar) [reply pkgs/os-specific/linux/prl-tools/default.nix:110]: Just to ensure -- does OpenGL work for AArch64 with this change? It seems strange that we skip creating sover symlinks for one architecture.

Comment (wegank) [reply pkgs/os-specific/linux/prl-tools/default.nix:110]: OpenGL support is provided by [virgl](https://kb.parallels.com/en/128518) starting from Parallels Desktop 17, so that it already works without Parallels Tools.

Comment (abbradar) [reply pkgs/os-specific/linux/prl-tools/default.nix:110]: I see, so am I correct that we don't need to create these symlinks at all nowadays, even on amd64? We already use kernel 5.10.

File: nixos/modules/virtualisation/parallels-guest.nix (hunk 5 above)
Comment (abbradar) [reply nixos/modules/virtualisation/parallels-guest.nix:70]: Same question about OpenGL here -- do we need to set `opengl.package` at all now?

Comment (wegank) [reply nixos/modules/virtualisation/parallels-guest.nix:70]: Seems that we should use the [VirGL 3D driver](https://virgil3d.github.io/) provided by Mesa instead.

Comment (abbradar) [reply nixos/modules/virtualisation/parallels-guest.nix:70]: Great! @wegank @paulsmith can you try commenting out all `hardware.opengl` stuff and verifying that OpenGL still works for you?

File: pkgs/os-specific/linux/prl-tools/default.nix (hunk 7 above)
Comment (wegank) [reply pkgs/os-specific/linux/prl-tools/default.nix:110]: Indeed. In this case, I think we should only copy ```lib/libPrl*.0.0``` in the install phase.

File: nixos/modules/virtualisation/parallels-guest.nix (hunk 5 above)
Comment (wegank) [reply nixos/modules/virtualisation/parallels-guest.nix:70]: @abbradar Yes, it works.

```
[nix-shell:~]$ glxinfo | grep OpenGL
OpenGL vendor string: Mesa/X.org
OpenGL renderer string: virgl
OpenGL core profile version string: 4.0 (Core Profile) Mesa 21.3.2
OpenGL core profile shading language version string: 4.00
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.1 Mesa 21.3.2
OpenGL shading language version string: 1.40
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.0 Mesa 21.3.2
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
OpenGL ES profile extensions:

[nix-shell:~]$ glxinfo | grep rendering
direct rendering: Yes
```

File: nixos/modules/virtualisation/parallels-guest.nix [hunk 11]
```diff
@@ -45,34 +47,36 @@ in
...
+          MatchIsPointer  "on"
+          MatchTag        "prlmouse"
+          Driver          "prlmouse"
         EndSection
       '';
+
+      screenSection = ''
+        Option "NoMTRR"
+      '';
     };
```
Comment (wegank) [inline nixos/modules/virtualisation/parallels-guest.nix:66]: All of these lines should be deleted: prlmouse and prlvideo no longer exist.

File: nixos/modules/virtualisation/parallels-guest.nix [hunk 12]
```diff
@@ -45,34 +47,36 @@ in
...
     services.udev.packages = [ prl-tools ];
 
     environment.systemPackages = [ prl-tools ];
 
     boot.extraModulePackages = [ prl-tools ];
 
-    boot.kernelModules = [ "prl_tg" "prl_eth" "prl_fs" "prl_fs_freeze" ];
+    boot.kernelModules = if aarch64 then [ "prl_fs" "prl_fs_freeze" "prl_notifier" "prl_tg" ] else [ "prl_fs" "prl_fs_freeze" "prl_tg" ];
 
     services.timesyncd.enable = false;
```
Comment (wegank) [inline nixos/modules/virtualisation/parallels-guest.nix:81]: Seems that this line is unnecessary also.

Comment (wegank) [reply nixos/modules/virtualisation/parallels-guest.nix:81]: Yes, it is an issue. We should definitely patch prltoolsd.

```
01-08 19:35:31.559 I /prlcc:1293:50d/ Running '/usr/bin/prldnd'
01-08 19:35:31.559 I /prlcc:1293:50d/ Running '/usr/bin/prlcp'
01-08 19:35:31.559 I /prlcc:1293:50d/ Running '/usr/bin/prlsga'
01-08 19:35:31.559 I /prlcc:1293:50d/ Running '/usr/bin/prlshprof'
01-08 19:35:31.559 F /prlcc:1299:513/ Error: failed to execute subprocess (errno 2)
```

**Comment by jamesla** [comment]
Is this good to merge?

**Comment by voanhduy1512** [comment]
I have been using this for last few months, it works great.

**Comment by Builditluc** [comment]
@voanhduy1512 can you please provide an example of how to use/apply this patch?

**Comment by jamesla** [comment]
@Builditluc @voanhduy1512 

Could you share how you got it working?

**Comment by Builditluc** [comment]
I used the modified files from @wegank and applied the patch from @paulsmith to enable clipboard sharing (found it in the issue, #31882, which led to this PR). [Here's](https://github.com/Builditluc/dotnix/tree/main/hosts/prlnix) the patch in my config.

**Comment by sandangel** [comment]
hi, could we merge this PR?

**Comment by joshpearce** [comment]
> I used the modified files from @wegank and applied the patch from @paulsmith to enable clipboard sharing (found it in the issue, #31882, which led to this PR). [Here's](https://github.com/Builditluc/dotnix/tree/main/hosts/prlnix) the patch in my config.

I'm using these files and having no issues.

File: nixos/modules/virtualisation/parallels-guest.nix [hunk 13]
```diff
@@ -4,6 +4,7 @@ with lib;
 
 let
   prl-tools = config.hardware.parallels.package;
+  aarch64 = pkgs.stdenv.hostPlatform.system == "aarch64-linux";
```
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/parallels-guest.nix:7]:
```suggestion
  isAarch64Linux = pkgs.stdenv.hostPlatform.system == "aarch64-linux";
```

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 14]
```diff
@@ -1,48 +1,55 @@
...
-    sha256 = "1jwzwif69qlhmfky9kigjaxpxfj0lyrl1iyrpqy4iwqvajdgbbym";
+    sha256 = "1ab5jwbg3jgvfwx8kwxwhhrsdp9wz4g9q7fg4z0hhd8v9pgy4yxx";
   };
 
   hardeningDisable = [ "pic" "format" ];
 
-  # also maybe python2 to generate xorg.conf
-  nativeBuildInputs = [ p7zip undmg ] ++ lib.optionals (!libsOnly) [ makeWrapper ] ++ kernel.moduleBuildDependencies;
+  nativeBuildInputs = [ p7zip undmg perl autoPatchelfHook ]
+    ++ lib.optionals (!libsOnly) [ makeWrapper ] ++ kernel.moduleBuildDependencies;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:31]:
```suggestion
  nativeBuildInputs = [ p7zip undmg perl autoPatchelfHook ]
    ++ kernel.moduleBuildDependencies
    ++ lib.optionals (!libsOnly) [ makeWrapper ];
```
To make clear that ``kernel.moduleBuildDependencies`` is not conditional.

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 15]
```diff
@@ -1,48 +1,55 @@
...
+
+  buildInputs = with xorg; [ stdenv.cc.cc libXrandr libXext libX11 libXcomposite libXinerama ]
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:33]:
```suggestion
  buildInputs = with xorg; [ libXrandr libXext libX11 libXcomposite libXinerama ]
```
We should have a compiler in the stdenv.

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 16]
```diff
@@ -1,48 +1,55 @@
...
+  runtimeDependencies = [ glib xorg.libXrandr ];
 
   inherit libsOnly;
 
   unpackPhase = ''
     undmg "${src}"
 
     export sourceRoot=prl-tools-build
-    7z x "Parallels Desktop.app/Contents/Resources/Tools/prl-tools-lin.iso" -o$sourceRoot
+    7z x "Parallels Desktop.app/Contents/Resources/Tools/prl-tools-lin${if aarch64 then "-arm" else ""}.iso" -o$sourceRoot
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:44]:
```suggestion
    7z x "Parallels Desktop.app/Contents/Resources/Tools/prl-tools-lin${lib.optionalString aarch64 "-arm"}.iso" -o$sourceRoot
```

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 17]
```diff
@@ -1,48 +1,55 @@
...
       ( cd $sourceRoot/kmods; tar -xaf prl_mod.tar.gz )
     fi
-    ( cd $sourceRoot/tools; tar -xaf prltools${if x64 then ".x64" else ""}.tar.gz )
+    # TODO
   '';
+    #( cd $sourceRoot/tools; tar -xaf prltools${if x64 then ".x64" else ""}.tar.gz )
 
-  kernelVersion = if libsOnly then "" else lib.getName kernel.name;
+  kernelVersion = if libsOnly then "" else lib.getVersion kernel.name;
   kernelDir = if libsOnly then "" else "${kernel.dev}/lib/modules/${kernelVersion}";
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:53]:
```suggestion
  kernelVersion = lib.optionalString (!libsOnly) lib.getVersion kernel.name;
  kernelDir = lib.optionalString (!libsOnly) "${kernel.dev}/lib/modules/${kernelVersion}";
```

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 18]
```diff
@@ -1,48 +1,55 @@
 { stdenv, lib, makeWrapper, p7zip
 , gawk, util-linux, xorg, glib, dbus-glib, zlib
 , kernel ? null, libsOnly ? false
-, undmg, fetchurl
+, fetchurl, undmg, perl, autoPatchelfHook
 }:
 
 assert (!libsOnly) -> kernel != null;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:7]:
```suggestion
```
kernel is never null under normal circumstances.

**Comment by sandangel** [comment]
I saw the parallel tools updated today. Will this patch still needed?

**Comment by sandangel** [comment]
I checked, the new parallel version doesn't need that patch. but it only works on kernel 5.17 . I tried compiling with kernel 5.18 but failed

**Comment by sandangel** [comment]
thanks for @wegank the prl-tools now updated and working with 17.1.3 and linux kernel 5.18

**Comment by Strum355** [comment]
With wegank's 17.1.4 prl-tools (not tested 17.1.3), Im getting the following output from the `prltoolsd` service, along with missing functionality such as no dynamic resolution scaling:
```
Jun 23 07:40:53 dev systemd[1]: Started Parallels Tools' service.
Jun 23 07:40:53 dev prltoolsd[715]: sh: line 1: awk: command not found
Jun 23 07:40:53 dev prltoolsd[737]: sh: line 1: /usr/bin/prlfsmountd: No such file or directory
Jun 23 07:40:53 dev prltoolsd[752]: sh: line 1: /usr/bin/prlfsmountd: No such file or directory
Jun 23 08:22:34 dev prltoolsd[9848]: sh: line 1: /usr/bin/prlfsmountd: No such file or directory
```

Adding `gawk` to the `prltoolsd` service's path only addresses the second log line, but functionality seems to be the same as before.

**Comment by catap** [comment]
I've created a PR: https://github.com/paulsmith/nixpkgs/pull/1 which
 - incorporated @wegank 's changes from https://github.com/wegank/nixos-config/tree/main/hardware/parallels-unfree
 - fixed a few remaining issues that @Strum355 mentored. 

I just tested it on Version 17.1.4 (51567) with on Intel mac and it works fine with OpenGL (see screenshot):

Screenshot 2022-06-27 at 12 46 35

**Comment by catap** [comment]
A kind of related PR: https://github.com/NixOS/nixpkgs/pull/179487

@wegank / @abbradar  / @paulsmith / @SuperSandro2000 is it possible to merge this one?

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 19]
```diff
@@ -57,112 +64,63 @@ stdenv.mkDerivation rec {
...
-
   meta = with lib; {
     description = "Parallels Tools for Linux guests";
     homepage = "https://parallels.com";
-    platforms = [ "i686-linux" "x86_64-linux" ];
+    platforms = [ "aarch64-linux" "i686-linux" "x86_64-linux" ];
     license = licenses.unfree;
-    # I was making this package blindly and requesting testing from the real user,
-    # so I can't even test it by myself and won't provide future updates.
-    maintainers = with maintainers; [ abbradar ];
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:166]: We must set a new maintainer.

File: pkgs/os-specific/linux/prl-tools/default.nix [hunk 20]
```diff
@@ -57,112 +64,63 @@ stdenv.mkDerivation rec {
...
   '';
 
-  dontStrip = true;
-  dontPatchELF = true;
-
   meta = with lib; {
     description = "Parallels Tools for Linux guests";
     homepage = "https://parallels.com";
-    platforms = [ "i686-linux" "x86_64-linux" ];
+    platforms = [ "aarch64-linux" "i686-linux" "x86_64-linux" ];
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/prl-tools/default.nix:123]:
```suggestion
    platforms = platform.linux;
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
There are still some review comments open which must be addressed.

**Comment by Strum355** [comment]
Out of curiosity, does dynamic resolution work for others on i3, or is it universally not working for any DE on NixOS?  (AFK so I cant provide prlcc logs right now)

File: pkgs/os-specific/linux/prl-tools/default.nix (hunk 20 above)
Comment (catap) [reply pkgs/os-specific/linux/prl-tools/default.nix:123]: @SuperSandro2000 I not sure that enable all linux which is supported by nix is good idea.

File: pkgs/os-specific/linux/prl-tools/default.nix (hunk 19 above)
Comment (catap) [reply pkgs/os-specific/linux/prl-tools/default.nix:166]: I can be a maintainer of this package, at least I'm quite motivated to keep it updated :)

File: pkgs/os-specific/linux/prl-tools/default.nix (hunk 20 above)
Comment (catap) [reply pkgs/os-specific/linux/prl-tools/default.nix:123]: from another hand it doesn't change anything. Just makes things clean.

**Comment by catap** [comment]
@SuperSandro2000 here the branch where I've addresed your comments: https://github.com/catap/nixpkgs/tree/prl-tools

File: pkgs/os-specific/linux/prl-tools/default.nix (hunk 19 above)
Comment (wegank) [reply pkgs/os-specific/linux/prl-tools/default.nix:166]: I can be a co-maintainer :)

**Comment by wegank** [comment]
@Strum355 It works for me on GNOME on Wayland.

File: pkgs/os-specific/linux/prl-tools/default.nix (hunk 19 above)
Comment (catap) [reply pkgs/os-specific/linux/prl-tools/default.nix:166]: @wegank just added also you :)

**Comment by catap** [comment]
@wegank / @SuperSandro2000 can I ask you to cherry-pick commits from my branch to this PR? Shall I open a new one and rebase it to the last master?

**Comment by dxmh** [comment]
> Out of curiosity, does dynamic resolution work for others on i3, or is it universally not working for any DE on NixOS? (AFK so I cant provide prlcc logs right now)

Dynamic resolution does not currently work for me with Xorg and dwm. (For what it’s worth, my workaround is to run `xrandr --output Virtual-1 --preferred`, which I have mapped to a keybinding.)

**Comment by Strum355** [comment]
@dxmh I was able to get a `udev` rule to do the work for me, maybe itll work for your purposes too:
```nix
{ pkgs, ... }:

let 
  xrandr-auto = (pkgs.writeShellScriptBin "xrandr-auto" ''
    ${pkgs.xorg.xrandr}/bin/xrandr --output Virtual-1 --preferred
  '');
in {
  # ...
  services.udev.extraRules = ''
    ACTION=="change", KERNEL=="card0", SUBSYSTEM=="drm", ENV{DISPLAY}=":0", ENV{XAUTHORITY}="/home/noah/.Xauthority", RUN+="${xrandr-auto}/bin/xrandr-auto" 
  '';
  # ...
}
```

**Comment by catap** [comment]
@Strum355 resize is working fine for me but I'm using unmerged PR :)

**Comment by Strum355** [comment]
@catap I was using your changes from 1-2 days ago ([diff here](https://github.com/paulsmith/nixpkgs/compare/d675eefdb0b4a0ad35caa5bd773b0dc01a3e72d9...1b5466e9b903167ce657cbfdcf5d92377a17b8c0)) but automatic resize wasn't working for me on i3wm.

**Comment by catap** [comment]
@Strum355 interesting, I've tested it with gnome and pantheon and it works.

**Comment by sandangel** [comment]
can we create another PR to nixpkgs master? It's hard to follow what changed on an unactive and closed PR

**Comment by catap** [comment]
@sandangel already did: https://github.com/NixOS/nixpkgs/pull/179582

**Comment by sandangel** [comment]
I guess we should move the conversation there then. thanks a lot for your fantastic work. @catap @wegank
